sparkly/testing.py

```diff
@@ -135,13 +172,294 @@ def assertDataFrameEqual(self, actual_df, expected_data, fields=None, ordered=Fa
+                    # ordering parameters
+                    ignore_order=False,
+                    ignore_order_depth=None,
+                    # float comparison parameters
+                    atol=0,
+                    rtol=1e-07,
+                    equal_nan=True,
+                    # DataType comparison parameters
+                    ignore_nullability=False):
+        """Assert equal on steroids.
+
+        Extend this classic function signature to work better with
+        comparisons involving rows, datatypes, dictionaries, lists and
+        floats by:
+            - ignoring the order of lists and datatypes recursively,
+            - comparing floats within a given tolerance,
+            - assuming NaNs are equal,
+            - ignoring the nullability requirements of datatypes (since
+              Spark can be inaccurate when inferring it),
+            - providing better diffs for rows and datatypes.
+
+        Float comparisons are inspired by NumPy's assert_allclose. The
+        main formula used is | float1 - float2 | <= atol + rtol * float2.
+
+        Args:
+            first, second, msg: see unittest.TestCase.assertEqual
+            ignore_order (bool|False): ignore the order in lists and
+                datatypes (rows, dicts are inherently orderless).
+            ignore_order_depth (int|None): if ignore_order is true, do
+                ignore order up to this level of nested lists or
+                datatypes (exclusive). Setting this to 0 or None means
+                ignore order infinitely, 1 means ignore order only at the
+                top level, 2 will ignore order within lists of lists and
+                so on. Default is ignore order arbitrarily deep.
+            atol (int, float|0): Absolute tolerance in float comparisons.
+            rtol (int, float|1e-07): Relative tolerance in float
+                comparisons.
+            equal_nan (bool|True): If set, NaNs will compare equal.
+            ignore_nullability (bool|False): If set, ignore all
+                nullability fields in dataTypes. This includes
+                containsNull in arrays, valueContainsNull in maps and
+                nullable in struct fields.
+
+        Returns:
+            None, iff the two objects are equal.
+
+        Raises
+            AssertionError, iff the two objects are not equal. See
+            unittest.TestCase.assertEqual for details.
+        """
+        # Our approach here is to redefine the 5 container objects that
+        # this function expects to work with - floats, dataTypes, rows,
+        # dicts and lists - to introduce generic ordering and extend
+        # the meaning of equality where applicable. We can then change
+        # all such objects to our custom containers, provide new asserters
+        # for some, and then feed them to the vanilla assertEqual.
+
+        # Define our custom containers
+        def cast_to_test_friendly_container(value, ignore_order_depth):
```

Comment:
did you consider to use http://pythonhosted.org/testfixtures/comparing.html#providing-your-own-comparers ? IMO, `textfixtures.compare` provides much more visibility on diff reports in case of complex structures.